core/src/main/java/io/grpc/internal/AbstractClientStream.java

```diff
@@ -119,8 +124,14 @@ protected void inboundHeadersReceived(Metadata.Headers headers) {
     }
     if (headers.containsKey(ChannelImpl.MESSAGE_ENCODING_KEY)) {
       String messageEncoding = headers.get(ChannelImpl.MESSAGE_ENCODING_KEY);
-      setDecompressor(messageEncoding);
+      try {
+        setDecompressor(messageEncoding);
+      } catch (IllegalArgumentException e) {
+        inboundTransportError(Status.INVALID_ARGUMENT.withDescription(e.getMessage()));
```

Comment:
Most status codes are directional. `INVALID_ARGUMENT` is defined as "Client specified an invalid argument." In this case the server specified something that the client doesn't support; that isn't `INVALID_ARGUMENT`.

Hmm... I'm thinking for now call both `sendCancel()` and `inboundTransportError()` and add a TODO to improve the situation. `sendCancel()` is delayed with Netty. I'm not confident about that solution, but everywhere I look I see potential problems, and that seems the least risky.